File: 2022/rust/src/bin/day07.rs
use aoc2022::Solve;
use std::io::{BufRead, BufReader, Read};

struct Solution {
    total_sizes: Vec<u32>,
}

impl Solve for Solution {
    type Answer1 = u32;
    type Answer2 = u32;

    fn new(r: impl Read) -> Self {
        let mut stack = Vec::new();
        let mut total_sizes = Vec::new();
        for line in BufReader::new(r).lines().filter_map(Result::ok) {
            match line.rsplit_once(char::is_whitespace) {
                Some(("$ cd", d)) => {
                    if d == ".." {
                        if let Some(total) = stack.pop() {
                            total_sizes.push(total);
                            if let Some(last) = stack.last_mut() {
                                *last += total;
                            }
                        }
                    } else {
                        stack.push(0);
                    }
                }
                Some((s, _)) => {
                    if let Ok(size) = s.parse::<u32>() {
                        if let Some(last) = stack.last_mut() {
                            *last += size;
                        }
                    }
                }
                _ => unreachable!(),
            }
        }
        while let Some(total) = stack.pop() {
            total_sizes.push(total);
            if let Some(last) = stack.last_mut() {
                *last += total;
            }
        }
        Self { total_sizes }
    }
    fn part1(&self) -> Self::Answer1 {
        self.total_sizes.iter().filter(|&x| *x <= 100_000).sum()
    }
    fn part2(&self) -> Self::Answer2 {
        *self
            .total_sizes
            .iter()
            .filter(|&x| self.total_sizes[self.total_sizes.len() - 1] - x < 40_000_000)
            .min()
            .unwrap()
    }
}

fn main() {
    let solution = Solution::new(std::io::stdin().lock());
    println!("Part 1: {}", solution.part1());
    println!("Part 2: {}", solution.part2());
}

#[cfg(test)]
mod tests {
    use super::*;

    fn example_input() -> &'static [u8] {
        r"
$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k
"[1..]
            .as_bytes()
    }

    #[test]
    fn example1() {
        assert_eq!(95437, Solution::new(example_input()).part1());
    }

    #[test]
    fn example2() {
        assert_eq!(24_933_642, Solution::new(example_input()).part2());
    }
}

File: 2022/python/aoc2022/day07.py
import sys
from collections import defaultdict
from typing import TextIO

from aoc2022 import Solve, run


class Solution(Solve):
    def __init__(self, io: TextIO) -> None:
        self.total_sizes: dict[tuple[str, ...], int] = defaultdict(int)
        curr: list[str] = []
        for line in io.readlines():
            parts = line.strip().split()
            if parts[:2] == ["$", "cd"]:
                if parts[2] == "..":
                    curr.pop()
                else:
                    curr.append(parts[2])
            if parts[0].isnumeric():
                for i in range(len(curr)):
                    self.total_sizes[tuple(curr[: i + 1])] += int(parts[0])

    def part1(self) -> int:
        return sum(filter(lambda x: x <= 100_000, self.total_sizes.values()))

    def part2(self) -> int:
        total = self.total_sizes[tuple(["/"])]
        return min(filter(lambda x: total - x < 40_000_000, self.total_sizes.values()))


if __name__ == "__main__":
    run(Solution(sys.stdin))
